### PR TITLE
[IOTDB-2848] Consensus snapshot interface

### PR DESCRIPTION
**Pull request by SzyWilliam**
## Description
In this PR, I add snapshot related interface to `IStateMachine`
Please refer to https://y8dp9fjm8f.feishu.cn/docs/doccnHeN9yVfAMgPAas6SAAaHTh# to see the design reason and usage for these interfaces.

### CODE REVIEW

File: consensus/src/main/java/org/apache/iotdb/consensus/statemachine/IStateMachine.java [hunk 1]
```diff
@@ -37,4 +40,12 @@ interface Registry extends Function<ConsensusGroupId, IStateMachine> {}
   TSStatus write(IConsensusRequest IConsensusRequest);
 
   DataSet read(IConsensusRequest IConsensusRequest);
+
+  void takeSnapshot(ByteBuffer metadata, File snapshotDir);
```
Comment (qiaojialin) [inline consensus/src/main/java/org/apache/iotdb/consensus/statemachine/IStateMachine.java:44]: add javadoc for method behavior and parameter

Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/statemachine/IStateMachine.java:57]: done

File: consensus/src/main/java/org/apache/iotdb/consensus/common/SnapshotMeta.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import java.io.File;
+import java.nio.ByteBuffer;
+import java.util.List;
+
+public class SnapshotMeta {
+  private ByteBuffer metadata;
+  private List<File> snapshotFiles;
+
+  public SnapshotMeta(ByteBuffer metadata, List<File> snapshotFiles) {
+    this.metadata = metadata;
```
Comment (wangchao316) [inline consensus/src/main/java/org/apache/iotdb/consensus/common/SnapshotMeta.java:30]: what does metadata contains ? add doc

File: consensus/src/main/java/org/apache/iotdb/consensus/statemachine/IStateMachine.java [hunk 3]
```diff
@@ -34,7 +37,50 @@ interface Registry extends Function<ConsensusGroupId, IStateMachine> {}
...
+   * read local data and return
+   *
+   * @param IConsensusRequest read request
+   */
   DataSet read(IConsensusRequest IConsensusRequest);
+
+  /**
+   * IConsensus will periodically take the snapshot on both log and statemachine Data
+   *
+   * @param metadata the metadata IConsensus want IStateMachine to preserve
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/statemachine/IStateMachine.java:57]: Javadoc states that the updated snapshot should have lexicographically larger metadata, which needs to be supported by various consensus algorithm implementations

Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/statemachine/IStateMachine.java:57]: done

File: consensus/src/main/java/org/apache/iotdb/consensus/common/SnapshotMeta.java (hunk 2 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/common/SnapshotMeta.java:30]: Done. In Ratis, metadata will contain ${term}_${index}. Other IConsensus implementation may vary. The guarantee here is that more updated snapshot will have lexicographically larger metadata.

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**